In [10]:
from surprise import Dataset
from surprise.model_selection import cross_validate
from surprise.prediction_algorithms import SVD, SVDpp, NMF



In [4]:
data = Dataset.load_builtin("ml-100k")

Dataset ml-100k could not be found. Do you want to download it? [Y/n] Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /Users/macair/.surprise_data/ml-100k


In [8]:
algo = SVD()
cross_validate(algo, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9305  0.9439  0.9353  0.9278  0.9363  0.9348  0.0055  
MAE (testset)     0.7340  0.7443  0.7383  0.7292  0.7395  0.7371  0.0051  
Fit time          0.73    0.66    0.65    0.76    0.66    0.69    0.05    
Test time         0.11    0.06    0.10    0.06    0.10    0.09    0.02    


{'test_rmse': array([0.93047443, 0.94386376, 0.9353255 , 0.92781303, 0.93630549]),
 'test_mae': array([0.73404793, 0.74425825, 0.73825489, 0.72923424, 0.73950042]),
 'fit_time': (0.7302720546722412,
  0.6575927734375,
  0.6491060256958008,
  0.764901876449585,
  0.6553840637207031),
 'test_time': (0.11314201354980469,
  0.0626060962677002,
  0.09954595565795898,
  0.0628809928894043,
  0.09501385688781738)}

In [11]:
algo_svdpp = SVDpp()
cross_validate(algo_svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9143  0.9269  0.9237  0.9115  0.9164  0.9186  0.0058  
MAE (testset)     0.7169  0.7307  0.7236  0.7167  0.7205  0.7217  0.0052  
Fit time          10.82   10.49   11.67   10.86   10.58   10.89   0.42    
Test time         2.26    2.25    2.25    2.27    2.20    2.25    0.02    


{'test_rmse': array([0.91425239, 0.92692029, 0.92373745, 0.91149875, 0.91643654]),
 'test_mae': array([0.71688253, 0.73068197, 0.72358223, 0.71666533, 0.72051645]),
 'fit_time': (10.822853088378906,
  10.494473934173584,
  11.671396970748901,
  10.856090068817139,
  10.581699848175049),
 'test_time': (2.259253978729248,
  2.253469944000244,
  2.254000186920166,
  2.2728257179260254,
  2.20158314704895)}

In [12]:
algo_nmf = NMF()
cross_validate(algo_nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9594  0.9620  0.9720  0.9722  0.9615  0.9654  0.0055  
MAE (testset)     0.7543  0.7569  0.7652  0.7624  0.7578  0.7593  0.0040  
Fit time          0.89    0.89    0.87    0.85    0.87    0.87    0.02    
Test time         0.05    0.09    0.05    0.09    0.05    0.07    0.02    


{'test_rmse': array([0.95937069, 0.96204093, 0.97202016, 0.97224724, 0.96146277]),
 'test_mae': array([0.75431653, 0.75685015, 0.76522551, 0.76244805, 0.75776894]),
 'fit_time': (0.890408992767334,
  0.8888771533966064,
  0.8654749393463135,
  0.8511440753936768,
  0.866102933883667),
 'test_time': (0.052127838134765625,
  0.09436798095703125,
  0.05368781089782715,
  0.09261798858642578,
  0.05249214172363281)}

Based on the evaluation results:

For SVD:
- Mean RMSE: 0.9348
- Mean MAE: 0.7371
- Mean Fit time: 0.69 seconds
- Mean Test time: 0.09 seconds

For SVDpp:
- Mean RMSE: 0.9186
- Mean MAE: 0.7217
- Mean Fit time: 10.89 seconds
- Mean Test time: 2.25 seconds

For NMF:
- Mean RMSE: 0.9654
- Mean MAE: 0.7593
- Mean Fit time: 0.87 seconds
- Mean Test time: 0.07 seconds

Based on RMSE and MAE, lower values are better. So, in this case, SVDpp seems to perform the best, as it has the lowest RMSE and MAE among the three algorithms. However, SVD also performs well and has significantly lower computational time compared to SVDpp. NMF has the highest RMSE and MAE among the three algorithms.

Therefore, if we prioritize accuracy, SVDpp is the best choice. But if computational efficiency is a concern, SVD might be preferred.